src/main.rs

```diff
@@ -66,47 +90,83 @@ fn main() -> Result<()> {
 
                 let rot = &mut camera.transform.rotation;
                 let new_rot = *rot * new_rot * rot.conj();
+
                 // Mathematically, the magnitude should always remain at 1 already, but floating point
                 // precision errors cause self-fueleing inaccuracy that becomes worse with each rotation.
                 let new_rot = new_rot * new_rot.mag().recip();
                 *rot = new_rot * *rot;
             };
 
-            match event.code {
-                KeyCode::Char('j') => rotation(std::f32::consts::FRAC_PI_8 / 4.0, Vec3::k()),
-                KeyCode::Char('l') => rotation(-std::f32::consts::FRAC_PI_8 / 4.0, Vec3::k()),
-                KeyCode::Char('k') => rotation(std::f32::consts::FRAC_PI_8 / 4.0, Vec3::j()),
-                KeyCode::Char('i') => rotation(-std::f32::consts::FRAC_PI_8 / 4.0, Vec3::j()),
-                KeyCode::Char('o') => rotation(std::f32::consts::FRAC_PI_8 / 4.0, Vec3::i()),
-                KeyCode::Char('u') => rotation(-std::f32::consts::FRAC_PI_8 / 4.0, Vec3::i()),
-                _ => (),
+            let mut did_rotation: bool = true;
+            if input.key_pressed(VirtualKeyCode::J) {
+                rotation(PI_FRAC_32, Vec3::k());
+            } else if input.key_pressed(VirtualKeyCode::L) {
+                rotation(-PI_FRAC_32, Vec3::k());
+            } else if input.key_pressed(VirtualKeyCode::K) {
+                rotation(PI_FRAC_32, Vec3::j());
+            } else if input.key_pressed(VirtualKeyCode::I) {
+                rotation(-PI_FRAC_32, Vec3::j());
+            } else if input.key_pressed(VirtualKeyCode::O) {
+                rotation(PI_FRAC_32, Vec3::i());
+            } else if input.key_pressed(VirtualKeyCode::O) {
+                rotation(-PI_FRAC_32, Vec3::i())
+            } else {
+                did_rotation = false;
             }
 
-            queue_render(&mut stdout, &world, &camera)?;
-            stdout.flush()?;
+            did_rotation || did_movement
+        } else {
+            false
+        };
+
+        // there is probably a way in the API to do this better but this was just very quick
+        let redraw_requested: bool = {
+            if let Event::RedrawRequested(_) = event {
+                true
+            } else {
+                false
+            }
+        };
```

Comment:
```suggestion
        let redraw_requested: bool = matches!(event, Event::RedrawRequested(_));
```

Reply:
Ah, I knew I could do something like this, I completely forgot how to at the time lol.

Reply:
Always remember to run `cargo clippy` (I even have it configured as rust-analyzer's checkOnSave command); clippy found this before I did.

Reply:
ok, I'll do that